### PR TITLE
Drainable stake manager

### PR DESCRIPTION
**Pull request by imyourm8**

### CODE REVIEW

File: contracts/common/misc/DrainStakeManager.sol [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import { StakeManagerStorage } from "../../staking/stakeManager/StakeManagerStorage.sol";
+import { Lockable } from "../mixin/Lockable.sol";
+
+contract DrainStakeManager is StakeManagerStorage {
+    constructor() public Lockable(address(0x0)) {}
+
+    function drain(address destination, uint amount) external onlyGovernance {
+        require(token.transfer(destination, amount));
+    }
+}
```
Comment (jdkanani) [inline contracts/common/misc/DrainStakeManager.sol:12]: Please add an empty line at the end of the file. You can set it automatically in vs code: https://stackoverflow.com/questions/44704968/visual-studio-code-insert-new-line-at-the-end-of-files

File: contracts/staking/stakeManager/IStakeManager.sol [hunk 2]
```diff
@@ -52,4 +50,7 @@ contract IStakeManager is StakeManagerStorage {
     // function lastStakedFor(address addr) external view returns (uint256);
     // function totalStakedForAt(address addr, uint256 blockNumber) external view returns (uint256);
     // function totalStakedAt(uint256 blockNumber) external view returns (uint256);
```
Comment (jdkanani) [inline contracts/staking/stakeManager/IStakeManager.sol:52]: Let's remove these functions if not required.

File: contracts/staking/stakeManager/StakeManager.sol [hunk 3]
```diff
@@ -44,6 +44,18 @@ contract StakeManager is IStakeManager {
         return NFTContract.ownerOf(tokenId);
     }
 
+    function epoch() public view returns(uint256) {
```
Comment (jdkanani) [inline contracts/staking/stakeManager/StakeManager.sol:47]: Any reason for adding a new function?

File: test/staking/DrainStakeManager.test.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+        it("must unlock stake manager", async function () {
+            await this.governance.update(
+                this.stakeManager.address,
+                this.stakeManager.contract.methods.unlock().encodeABI()
+            );
+            (await this.stakeManager.locked()).should.be.equal(false);
+        });
+    });
+});
```
Comment (jdkanani) [inline test/staking/DrainStakeManager.test.js:103]: Please add new line at end of the file.

File: contracts/common/misc/DrainStakeManager.sol (hunk 1 above)
Comment (jdkanani) [reply contracts/common/misc/DrainStakeManager.sol:12]: The reasoning behind it: https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file

File: contracts/staking/stakeManager/StakeManager.sol (hunk 3 above)
Comment (coinsandsteeldev) [reply contracts/staking/stakeManager/StakeManager.sol:47]: ValidatorShare.sol uses currentEpoch getter which is public getter from StakeManagerStorage. After refactor is no longer presented in IStakeManager and i had to add for compilation reasons.

File: contracts/common/misc/DrainStakeManager.sol [hunk 5]
```diff
@@ -0,0 +1,12 @@
+pragma solidity ^0.5.2;
+
+import { StakeManagerStorage } from "../../staking/stakeManager/StakeManagerStorage.sol";
+import { Lockable } from "../mixin/Lockable.sol";
+
+contract DrainStakeManager is StakeManagerStorage {
+    constructor() public Lockable(address(0x0)) {}
+
+    function drain(address destination, uint amount) external onlyGovernance {
+        require(token.transfer(destination, amount));
```
Comment (atvanguard) [inline contracts/common/misc/DrainStakeManager.sol:10]: It's a good practice to add a revert message as a second param to `require`.

File: contracts/staking/stakeManager/IStakeManager.sol [hunk 6]
```diff
@@ -1,8 +1,6 @@
 pragma solidity ^0.5.2;
 
-import {StakeManagerStorage} from "./StakeManagerStorage.sol";
-
-contract IStakeManager is StakeManagerStorage {
```
Comment (atvanguard) [inline contracts/staking/stakeManager/IStakeManager.sol:5]: Many thanks. ⭐

File: contracts/staking/stakeManager/StakeManager.sol (hunk 3 above)
Comment (atvanguard) [reply contracts/staking/stakeManager/StakeManager.sol:47]: Nice!

File: test/helpers/artifacts.js [hunk 7]
```diff
@@ -1,50 +1,51 @@
...
-export const childContracts = {
-  Marketplace: artifacts.require('Marketplace'),
-  ChildChain: artifacts.require('ChildChain'),
-  ChildERC20: artifacts.require('ChildERC20'),
-  ChildERC721: artifacts.require('ChildERC721'),
-  ChildERC721Mintable: artifacts.require('ChildERC721Mintable'),
-  MRC20: artifacts.require('MRC20'),
-  TestMRC20: artifacts.require('TestMRC20')
-}
+export const RootChain = artifacts.require('RootChain')
```
Comment (atvanguard) [inline test/helpers/artifacts.js:1]: All these lines are both red and green, is this a whitespace change? Also in `DepositManager.test.js`

File: test/helpers/deployer.js [hunk 8]
```diff
@@ -104,7 +104,10 @@ class Deployer {
       registry: this.registry,
       rootChain: this.rootChain,
       stakeManager: this.stakeManager,
-      stakeToken: this.stakeToken
+      stakeToken: this.stakeToken,
+      governance: this.governance,
+      proxy: proxy,
```
Comment (atvanguard) [inline test/helpers/deployer.js:109]: How about giving it a more descriptive name?

File: contracts/common/misc/DrainStakeManager.sol (hunk 5 above)
Comment (coinsandsteeldev) [reply contracts/common/misc/DrainStakeManager.sol]: For sure, in this particular place i omitted this message for not being informative enough. I added it anyway to follow best practices.

File: test/helpers/artifacts.js (hunk 7 above)
Comment (coinsandsteeldev) [reply test/helpers/artifacts.js:1]: image
It seems like, i'm not sure what "<<" represents here (it is new addition in my file).

File: test/helpers/deployer.js (hunk 8 above)
Comment (coinsandsteeldev) [reply test/helpers/deployer.js:109]: It follows project's terms and contract names. Any suggestions?

Comment (coinsandsteeldev) [reply test/helpers/deployer.js:109]: I renamed proxy -> stakeManagerProxy to be more clear what kind of proxy is that.

**Pull request merged**